services/types.go

```diff
@@ -248,6 +250,108 @@ type PlacementOptions interface {
+	ClockOptions() clock.Options
+
+	// SetInstrumentOptions sets the instrument options.
+	SetInstrumentOptions(value instrument.Options) StagedPlacementWatcherOptions
+
+	// InstrumentOptions returns the instrument options.
+	InstrumentOptions() instrument.Options
+
+	// SetActiveStagedPlacementOptions sets the active staged placement options.
+	SetActiveStagedPlacementOptions(value ActiveStagedPlacementOptions) StagedPlacementWatcherOptions
+
+	// ActiveStagedPlacementOptions returns the active staged placement options.
+	ActiveStagedPlacementOptions() ActiveStagedPlacementOptions
+
+	// SetStagedPlacementKey sets the kv key to watch for staged placement.
+	SetStagedPlacementKey(value string) StagedPlacementWatcherOptions
+
+	// StagedPlacementKey returns the kv key to watch for staged placement.
+	StagedPlacementKey() string
+
+	// SetStagedPlacementStore sets the staged placement store.
+	SetStagedPlacementStore(store kv.Store) StagedPlacementWatcherOptions
+
+	// StagedPlacementStore returns the staged placement store.
+	StagedPlacementStore() kv.Store
+
+	// SetInitWatchTimeout sets the initial watch timeout.
+	SetInitWatchTimeout(value time.Duration) StagedPlacementWatcherOptions
+
+	// InitWatchTimeout returns the initial watch timeout.
+	InitWatchTimeout() time.Duration
+}
+
+// ActiveStagedPlacement describes active staged placement.
+type ActiveStagedPlacement interface {
+	// ActivePlacement returns the currently active placement for a given time, the callback
+	// function when the caller is done using the placement, and any errors encountered.
+	ActivePlacement() (ServicePlacement, DoneFn, error)
+
+	// Close closes the active staged placement.
+	Close() error
+}
+
+// OnPlacementsAddedFn is called when placements are added.
+type OnPlacementsAddedFn func(placements []ServicePlacement)
+
+// OnPlacementsRemovedFn is called when placements are removed.
+type OnPlacementsRemovedFn func(placements []ServicePlacement)
+
+// ActiveStagedPlacementOptions provide a set of options for active staged placement.
+type ActiveStagedPlacementOptions interface {
+	// SetClockOptions sets the clock options.
+	SetClockOptions(value clock.Options) ActiveStagedPlacementOptions
+
+	// ClockOptions returns the clock options.
+	ClockOptions() clock.Options
+
+	// SetOnPlacementsAddedFn sets the callback function for adding placement.
+	SetOnPlacementsAddedFn(value OnPlacementsAddedFn) ActiveStagedPlacementOptions
+
+	// OnPlacementsAddedFn returns the callback function for adding placement.
+	OnPlacementsAddedFn() OnPlacementsAddedFn
+
+	// SetOnPlacementsRemovedFn sets the callback function for removing placement.
+	SetOnPlacementsRemovedFn(value OnPlacementsRemovedFn) ActiveStagedPlacementOptions
+
+	// OnPlacementsRemovedFn returns the callback function for removing placement.
+	OnPlacementsRemovedFn() OnPlacementsRemovedFn
+}
+
+// StagedPlacement describes a series of placements applied in staged fashion.
+type StagedPlacement interface {
```

Comment:
Not for now, although I guess yeah if we want to generate a new placement and persist in kv, sure can add.